### ISSUE TITLE
Unable to add to package.searchers with class instance

### ISSUE THREAD

**Issue by pcsagan** [ThePhD/sol2#1172]
I have a Lua class that holds onto a sol::state, and it has a function for handling missing require modules. If I pass in a static function and no this pointer, then it works fine, but if I pass in a class method with a this pointer, I get the following error:

sol: runtime error: stack index 1, expected userdata, received string: value is not a valid userdata (bad argument into 'int(lua_State*)')
stack traceback:
	[C]: in ?
	[C]: in global 'require'
	[strin...

I'm not sure what's causing the exception text to be truncated - that's straight out of e.what().

This is what I'm doing to add the path to package.searchers:

```cpp
state_.open_libraries(sol::lib::table);
state_.open_libraries(sol::lib::package);
state_.set_function("my_loader", &Lua::RequireModule, this);
state_.script("table.insert(package.searchers, my_loader)");
```

**Comment by ThePhD** [comment]
What is the signature of `RequireModule`, and what is the class it's in?

**Comment by pcsagan** [comment]
Signature:
```cpp
int RequireModule(lua_State* l)
```

Header:
```cpp
#pragma once

#include <Urho3D/Core/Object.h>
#include <Urho3D/Core/Context.h>
#include <sol/state.hpp>

class Lua : public Urho3D::Object
{
    URHO3D_OBJECT(Lua, Urho3D::Object);
public:
    explicit Lua(Urho3D::Context* context);
    virtual ~Lua();

    static void RegisterObject(Urho3D::Context* context);

    int RequireModule(lua_State* l)
    {
        std::string moduleName = lua_tostring(l, 1);
        return 1;
    }

    sol::state state_;
};
```

Source:
```cpp
#include "Lua.h"

using namespace Urho3D;

Lua::Lua(Context* context) :
    Object(context)
{    
    try
    {
        state_.open_libraries(sol::lib::table);
        state_.open_libraries(sol::lib::package);
        state_.set_function("my_loader", &Lua::RequireModule, this);
        state_.script("table.insert(package.searchers, my_loader)");
    }
    catch (sol::error& e)
    {
        std::string test = e.what();
    }
}

Lua::~Lua()
{
}

void Lua::RegisterObject(Context* context)
{
    context->RegisterSubsystem<Lua>();
}
```

**Comment by ThePhD** [comment]
Yes, this code won't work. You cannot bind a "raw Lua function, but also it's a member function": there's no such thing. Rework the function signature:

```cpp
    sol::table RequireModule(sol::state_view Lview, std::string_view moduleName)
    {
        lua_State* L = Lview;
        // create the module as a Lua thing, return it here
        sol::table my_module = state_.create_table(L);
        // fill with stuff
        return my_module;
    }
```

You can return anything you want, we're returning a `sol::table` because that's the most common.

Good luck!

**Comment by pcsagan** [comment]
Sorry, but could you clarify whether or not it's possible to bind a class member function to Lua in sol2? You appear to have changed the signature, but you also appear to have not placed it within a class, and you are not suggesting to pass the this pointer as a third parameter to set_function().

Did you mean to suggest that it's not possible at all to bind to a class member function, and you merely also suggested to modify the already-working "raw" Lua function to a more modern variant? Or did you mean to suggest that modifying the class member function's signature to what you suggested, then passing the this pointer as a third parameter to set_function(), is valid?

I tried modifying my class member to match the signature you suggested, and continued to call ```cpp state_.set_function("my_loader", &Lua::RequireModule, this);```, but the result was the same as before: expected userdata, received string.

The documentation seems to suggest that the third parameter of set_function() can be an object instance, but at the same time the example uses some wonky struct and doesn't actually pass in a pointer to an object, so maybe I misread the docs?

**Comment by ThePhD** [comment]
Oh, I see. Sorry, I had made the wrong assumptions for what you were trying to do. There's also an error in my code (it should be `sol::this_state`, not `sol::state_view`).

Let me start from the top:

# Yes, you can use Member Functions like this.

Here is a full example. I don't have Urho3D so right now I just put some "stubs" in that don't mean anything:

```cpp
#define SOL_ALL_SAFETIES_ON 1
#include <sol/sol.hpp>

namespace Urho3D {
	class Context;

	class Object {
	public:
		Object(Context*) noexcept {
		}
	};
} // namespace Urho3D

#define URHO3D_OBJECT(...)

class Lua : public Urho3D::Object {
	URHO3D_OBJECT(Lua, Urho3D::Object);

public:
	explicit Lua(Urho3D::Context* context);
	virtual ~Lua();

	static void RegisterObject(Urho3D::Context* context);

	sol::table LoadTestModule() {
		// create the module as a Lua thing, return it here
		sol::table my_module = state_.create_table();
		// fill with stuff
		my_module["name"] = "test";
		my_module["hi"] = "hello";
		return my_module;
	}

	sol::object RequireModule(std::string_view moduleName) {
		if (moduleName != "test") {
			// not the module we're looking for
			return sol::lua_nil;
		}
		// this is the test module, and we will return a
		// package loader function for it
		sol::object my_load_function = sol::make_object(state_, sol::as_function<sol::table (Lua::*)()>(&Lua::LoadTestModule, this));
		return my_load_function;
	}

	sol::state state_;
};

using namespace Urho3D;

Lua::Lua(Context* context) : Object(context) {
	try {
		state_.open_libraries(sol::lib::table, sol::lib::package, sol::lib::base);
		state_.set_function("my_loader", &Lua::RequireModule, this);
		state_.script("table.insert(package.searchers, my_loader)");
	}
	catch (sol::error& e) {
		std::string test = e.what();
	}
}

Lua::~Lua() {
}

void Lua::RegisterObject(Context*) {
	// context->RegisterSubsystem<Lua>();
}

int main() {

	Urho3D::Context* context = nullptr;

	Lua lua(context);

	lua.state_.script(R"(
		local test_mod = require("test")
		print(test_mod.name)
		print(test_mod.hi)
	)");

	return 0;
}
```

Passing in a pointer (`this` or otherwise) works fine. The example above creates a Module Loader Function, and then the actual Module Loading function. It should do the job. Note that we use the special syntax `sol::as_function` when we want to create an "anonymous" function that is not specifically tied to a key, like `"my_loader"`.

Does that clear things up?

**Comment by pcsagan** [comment]
This is crystal clear, and it works great.

For the record I'm using a fork of Uhro3D called rbfx, which uses EA's STL library, which lead me down a path of supporting custom string types, which lead me to https://sol2.readthedocs.io/en/latest/tutorial/customization.html where I believe A) the non-boilerplate code should be in the sol namespace, and B) the sol_lua_push() function is missing sol::types<two_things> as a first parameter. The errors and crashes I encountered were, at least to me, not indicative at all of what was actually wrong (it got a little crazy before I realized what was happening). Also if C++20 preview features are turned on lambdas no longer play well with set_function. Anyway, thanks for the great work!
